### ISSUE TITLE
Poor ATMForce() performance due to lack of atom reordering

### ISSUE THREAD

**Issue by egallicc** [openmm/openmm#4395]
I am investigating poor ATMForce performance for the common case where `ATMForce()` is given a NonbondedForce that is removed from the System. 

I tracked the issue down to ComputeContext skipping `reorderAtomsImpl()` because `getNonbondedUtilities().getUseCutoff()` is `false` (in addition, `getNonbondedUtilities().getUsePeriodic()` is also `false`). I verified that the performance improves by a factor of two by forcing atom reordering.

Based on my tests, `ComputeContext::reorderAtoms()` is only called in the outer Context. Does this reflect the design? If so, it would make sense that reordering is not carried out because the outer context does not manage a NonbondedForce.

Assuming that this analysis is correct, is there a clean way to force atom reordering in the outer Context even when it does not manage a NonbondedForce?

**Comment by egallicc** [comment]
We found a [workaround](https://github.com/Gallicchio-Lab/AToM-OpenMM/commit/be867de26bbf638576dc8c7db07167abf0f7e64a) for AToM-OpenMM that disables the NonbondedForce using force groups rather than removing it from the outer System. As a result, atom reordering appears to be triggered correctly resulting in a ~2x speed-up for protein-sized systems. This setup is similar to the one we had in place with the ATMMetaForce() plugin.

**Comment by peastman** [comment]
I'm confused about how the code works, especially the ReorderListener:

https://github.com/openmm/openmm/blob/71b2a93e09d9d5f110310448c9ff503eb2e71d55/platforms/common/src/CommonKernels.cpp#L8154-L8164

It seems to be forcing the inner context to work with unsorted coordinates?  That will be very inefficient.

Compare to CustomCVForce which lets the inner context order the particles however it wants and just maintains a mapping between them for use when copying data:

https://github.com/openmm/openmm/blob/71b2a93e09d9d5f110310448c9ff503eb2e71d55/platforms/common/src/CommonKernels.cpp#L5443-L5449

**Comment by egallicc** [comment]
I also have been doubtful about the current strategy. However, I tried the reordering strategy in CustomCVForce() and did not resolve the poor performance. After some testing, I almost convinced myself that the atom reordering routine is never called in the inner Contexts. Is this possible? If so, it would explain why the different strategies lead to the same outcome.

ATMForce()'s copyState() function:

https://github.com/openmm/openmm/blob/71b2a93e09d9d5f110310448c9ff503eb2e71d55/platforms/common/src/kernels/atmforce.cc#L15-L40 

transfers the coordinates to the inner Contexts without taking into account atom ordering. I think this is equivalent to imposing the outer Context's ordering to the inner Contexts.

Based on this analysis, the poor performance we have been experiencing was due to the outer Context skipping atom ordering when no nonbondedForce is present in the outer System.

In summary, it would be great to get your opinion on:

- The possibility that the inner contexts skip atom reordering
- The logic of transferring coordinates from the outer to the inner contexts without taking into account atom reordering. It seems to me that doing that is equivalent to assigning the same index to molecules with the same coordinates.

**Comment by peastman** [comment]
I think you're right.  The integrator calls `reorderAtoms()` on every time step, but that only causes the main context to be reordered.  Nothing ever calls it on the inner context.

**Comment by egallicc** [comment]
Right. Our [workaround](https://github.com/Gallicchio-Lab/AToM-OpenMM/commit/be867de26bbf638576dc8c7db07167abf0f7e64a) is to enable reordering in the main Context by retaining the NonbondedForce in the System, but skipping its evaluation by placing into a force group which is not evaluated.

The ordering in the main Context is then propagated to the inner Contexts implicitly by transferring coordinates from the outer to the inner contexts without taking into account the reordered atom indexes, as described above. I have added a test to probe this NonbondedForce setup, which passes. It's coming up in a PR. We have also done extensive simulations with AToM-OpenMM to check that the approach is correct. But please let me know if you see any potential pitfalls.

**Comment by peastman** [comment]
I believe that should work fine as a workaround.  But we should modify the code so it isn't needed.  The inner context should be able to reorder atoms as needed to efficiently evaluate its forces.

I'll tag this issue as a bug.

**Comment by egallicc** [comment]
Agreed. Here are some potential issues to think about:

- If the inner contexts have only a NonbondedForce, they might need the full set of ForceInfo's from the main Context to correctly recognize and reorder molecules.
- ATMForce() uses a displacement in lab coordinates that is unaware of periodic boundary conditions.  Unless they are grouped in the same "molecule", the ligand and receptor's positions could be PBC-wrapped differently, potentially upsetting their relative displacement. AToM-OpenMM includes a restraint that tethers the ligand to the binding site of the receptor, making them one molecule. But it is probably an issue to keep in mind for general applications of ATMForce() with a NonbondedForce.

**Comment by peastman** [comment]
> If the inner contexts have only a NonbondedForce, they might need the full set of ForceInfo's from the main Context to correctly recognize and reorder molecules.

I don't think that's an issue.  If each context does its own reordering and we swap orders correctly when copying between them (that's how CustomCVForce works), it doesn't matter if they're based on different forces.  Each context will pick an order that gives correct results for the forces it contains.
